tests/test_minilinq.py

```diff
@@ -88,6 +85,33 @@ def test_eval_auto_id_reference_nested(self):
         #   Reference('$.foo.id'):
         #       '1.bid' -> 'bid'
 
+    def test_flatmap_value_or_root(self):
+        """Low level test case for 'value-or-root' use case"""
+        env = BuiltInEnv() | JsonPathEnv({})
+
+        data = [
+            {"id": 1, "foo": {'id': 'bid', 'name': 'zip'}, "bar": [{'baz': 'a1'}, {'baz': 'a2', 'id': 'bazzer'}]},
+            {"id": 2, "foo": {'id': 'bid', 'name': 'zap'}, "bar": []},
+            {"id": 3, "foo": {'id': 'bid', 'name': 'mip'}, "bar": {}},
+            # {"id": 4, "foo": {'id': 'bid', 'name': 'map'}, "bar": None},  # fails with TypeError from jsonpath
+            {"id": 5, "foo": {'id': 'bid', 'name': 'mop'}},
+        ]
+        value_or_root = Apply(
+            Reference('_or_raw'), Reference(str('bar.[*]')), Reference("$")
+        )
+        flatmap = FlatMap(source=Literal(data), body=value_or_root)
+        mmap = Map(source=flatmap, body=List([
+            Reference("id"), Reference('baz'), Reference('$.id'), Reference('$.foo.id'), Reference('$.foo.name')
+        ]))
+        self.check_case(mmap.eval(env), [
+            ['1.bar.1.bar.[0]', 'a1', '1', '1.bid', 'zip'],
+            ['1.bar.bazzer', 'a2', '1', '1.bid', 'zip'],
+            ['2', [], '2', '2.bid', 'zap'],
+            ['3.bar.3.bar.[0]', [], '3', '3.bid', 'mip'],
```

Comment:
(also find it odd how different the `id` is in these two examples. it's because the dict counts as an entry but an empty list does not?)

Reply:
your confusion is well deserved. There are some 'bugs' and oddities in commcare-export as well as jsonpath-rw that cause this. I pushed a test to master earlier in the day to 'document' the weirdness: https://github.com/dimagi/commcare-export/blob/master/tests/test_minilinq.py#L58

I am planning to replace jsonpath-rw with jsonpath-ng which has a bunch of bugfixes and improvements and will resolve this specific issue (for this case 'id' will become '3').

Reply:
Ah, thanks I missed that. Sounds like I can mostly ignore the first column in these then? If so then everything is much clearer.

Reply:
(could we remove the id column from this test if it's already tested elsewhere and not part of what's being tested here?)

Reply:
https://github.com/dimagi/commcare-export/pull/190